### PR TITLE
Add group membership management

### PR DESCRIPTION
**Pull request by tiran**
A group membership manager is a user or a group that can add members to
a group or remove members from a group.

Fixes: https://pagure.io/freeipa/issue/8114
Signed-off-by: Christian Heimes <cheimes@redhat.com>

### CODE REVIEW

File: install/share/60basev2.ldif [hunk 1]
```diff
@@ -21,7 +21,9 @@ objectClasses: (2.16.840.1.113730.3.8.4.14 NAME 'ipaEntitlement' DESC 'IPA Entit
...
 objectClasses: (2.16.840.1.113730.3.8.4.2 NAME 'ipaService' DESC 'IPA service objectclass' AUXILIARY MAY ( memberOf $ managedBy $ ipaKrbAuthzData) X-ORIGIN 'IPA v2' )
 objectClasses: (2.16.840.1.113730.3.8.4.3 NAME 'nestedGroup' DESC 'Group that supports nesting' SUP groupOfNames STRUCTURAL MAY memberOf X-ORIGIN 'IPA v2' )
-objectClasses: (2.16.840.1.113730.3.8.4.4 NAME 'ipaUserGroup' DESC 'IPA user group object class' SUP nestedGroup STRUCTURAL X-ORIGIN 'IPA v2' )
+# Same for memberManager except it's actually v4 attribute
+attributeTypes: (2.16.840.1.113730.3.8.3.28 NAME 'memberManager' DESC 'DNs of entries allowed to manage group membership' SUP distinguishedName EQUALITY distinguishedNameMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.12 X-ORIGIN 'IPA v4')
+objectClasses: (2.16.840.1.113730.3.8.4.4 NAME 'ipaUserGroup' DESC 'IPA user group object class' SUP nestedGroup STRUCTURAL MAY memberManager X-ORIGIN 'IPA v2' )
 objectClasses: (2.16.840.1.113730.3.8.4.5 NAME 'ipaHostGroup' DESC 'IPA host group object class' SUP nestedGroup STRUCTURAL X-ORIGIN 'IPA v2' )
```
Comment (simo5) [inline install/share/60basev2.ldif]: add to ipaHostGroup too (including UI machinery), so that we can let people administer groups of hosts too

Comment (tiran) [reply install/share/60basev2.ldif]: The last version has memberManager for host groups, too.

File: doc/designs/membermanager.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+## Implementation
+
+The user group commands and host group commands are extended to handle
+member managers. The plugin classes grow two additional sub commands,
+one for adding and one for removing member managers. The show command
+prints member manager users and member manager groups. The find command
+can search by member manager.
+
+Member managers are stored in a new LDAP attribute ``memberManager``
+which is a multi-valued DN. The attribute can be added to entries with
```
Comment (abbra) [inline doc/designs/membermanager.md]: "a multi-valued attribute that contains DNs of objects which can manage members of the group"

File: install/share/60basev2.ldif [hunk 3]
```diff
@@ -21,8 +21,10 @@ objectClasses: (2.16.840.1.113730.3.8.4.14 NAME 'ipaEntitlement' DESC 'IPA Entit
 objectClasses: (2.16.840.1.113730.3.8.4.15 NAME 'ipaPermission' DESC 'IPA Permission objectclass' AUXILIARY MAY ( ipaPermissionType ) X-ORIGIN 'IPA v2' )
 objectClasses: (2.16.840.1.113730.3.8.4.2 NAME 'ipaService' DESC 'IPA service objectclass' AUXILIARY MAY ( memberOf $ managedBy $ ipaKrbAuthzData) X-ORIGIN 'IPA v2' )
 objectClasses: (2.16.840.1.113730.3.8.4.3 NAME 'nestedGroup' DESC 'Group that supports nesting' SUP groupOfNames STRUCTURAL MAY memberOf X-ORIGIN 'IPA v2' )
-objectClasses: (2.16.840.1.113730.3.8.4.4 NAME 'ipaUserGroup' DESC 'IPA user group object class' SUP nestedGroup STRUCTURAL X-ORIGIN 'IPA v2' )
-objectClasses: (2.16.840.1.113730.3.8.4.5 NAME 'ipaHostGroup' DESC 'IPA host group object class' SUP nestedGroup STRUCTURAL X-ORIGIN 'IPA v2' )
+# Same for memberManager except it's actually v4 attribute
+attributeTypes: (2.16.840.1.113730.3.8.3.28 NAME 'memberManager' DESC 'DNs of entries allowed to manage group membership' SUP distinguishedName EQUALITY distinguishedNameMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.12 X-ORIGIN 'IPA v4')
```
Comment (abbra) [inline install/share/60basev2.ldif]: Please add the attribute to RHANANA repository if not done yet.

File: doc/designs/membermanager.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+## Implementation
+
+The user group commands and host group commands are extended to handle
+member managers. The plugin classes grow two additional sub commands,
+one for adding and one for removing member managers. The show command
+prints member manager users and member manager groups. The find command
+can search by member manager.
+
+Member managers are stored in a new LDAP attribute ``memberManager``
+with OID 2.16.840.1.113730.3.8.3.28. It is multi-valued and contains
```
Comment (abbra) [inline doc/designs/membermanager.md]: As discussed on IRC, I moved the attribute into a new OID space for v4 attributes. Please use `2.16.840.1.113730.3.8.23.1.1` for `memberManager`

**Comment by abbra** [comment]
LGTM.

**Comment by tiran** [comment]
master:

* f0a1f084b64bd0ca570999594f8e25d4e8808938 Add group membership management
* 0f4c41ab2609171029e0a4334585688fc7c054f2 Add tests for member management

**Pull request closed without merging**